subsys/nfc/tnep/tnep.c

```diff
@@ -0,0 +1,576 @@
+		LOG_DBG("period timeout");
+	break;
+	case TNEP_EVENT_N_WAIT_TIMEOUT:
+		LOG_DBG("communication timeout!");
+
+		last_service->callbacks->timeout();
+	break;
+	default:
+		LOG_ERR("unknown signal %d", event);
+	break;
+	}
+
+	tnep_error_check(err);
+}
+
+const struct tnep_state_machine tnep_states[] = {
+	{ TNEP_STATE_DISABLED, tnep_sm_disabled },
+	{ TNEP_STATE_SERVICE_READY, tnep_sm_service_ready },
+	{ TNEP_STATE_SERVICE_SELECTED, tnep_sm_service_selected },
+};
+
+int nfc_tnep_set_message_buffer(u8_t *buffer, size_t buffer_length)
+{
+	if (!buffer || !buffer_length) {
+		return -EINVAL;
+	}
+
+	tnep_msg_buffer_size = buffer_length;
+	tnep_msg_buffer = buffer;
+
+	return 0;
+}
+
+
+void nfc_tnep_new_message_indicator(void)
+{
+	k_poll_signal_raise(&msg_signal, TNEP_SIGNAL_MESSAGE_RECIVED);
+
+	if (active_service) {
+		nfc_tnep_timer_start();
+	}
+}
+
+int nfc_tnep_init(struct nfc_tnep_service *services, size_t services_amount)
+{
+	LOG_INF("tnep init");
+
+	int err = 0;
+
+	if (current_state != TNEP_STATE_DISABLED) {
+		return -ENOTSUP;
+	}
+
+	LOG_DBG("k_pool init");
+	k_poll_signal_init(&msg_signal);
+	k_poll_signal_init(&timer_signal);
+
+	k_poll_event_init(&nfc_tnep_events[0],
+			  K_POLL_TYPE_SIGNAL,
+			  K_POLL_MODE_NOTIFY_ONLY,
+			  &msg_signal);
+
+	k_poll_event_init(&nfc_tnep_events[1],
+			  K_POLL_TYPE_SIGNAL,
+			  K_POLL_MODE_NOTIFY_ONLY,
+			  &timer_signal);
+
+	LOG_DBG("service init");
+	if (!services_amount || !services) {
+		return -EINVAL;
+	}
+
+	LOG_DBG("output buffer check");
+	if (!tnep_msg_buffer || !tnep_msg_buffer_size) {
+		return -EIO;
+	}
+
+	tnep_services = services;
+	tnep_services_amount = services_amount;
+
+	LOG_DBG("ndef message init");
+	err = tnep_message_add_records_services();
+
+	if (err) {
+		LOG_ERR("Error adding service record. Code: %d", err);
+		return err;
+	}
+
+	LOG_DBG("state machine init");
+	current_state = TNEP_STATE_SERVICE_READY;
+
+	k_poll_signal_raise(&msg_signal, TNEP_SIGNAL_MESSAGE_INIT);
+
+	return 0;
+}
+
+int nfc_tnep_process(void)
+{
+	int err = k_poll(nfc_tnep_events, NFC_TNEP_EVENTS_NUMBER, 100);
+
+	if (err) {
+		return err;
+	}
+
+	if (nfc_tnep_events[0].state == K_POLL_STATE_SIGNALED) {
+		nfc_tnep_events[0].signal->signaled = 0;
+		nfc_tnep_events[0].state = K_POLL_STATE_NOT_READY;
+
+		tnep_states[current_state].action(TNEP_EVENT_MESSAGE_OFFSET
+					+ nfc_tnep_events[0].signal->result);
+
+		tnep_message_encode();
+
+	} else if (nfc_tnep_events[1].state == K_POLL_STATE_SIGNALED) {
+		nfc_tnep_events[1].signal->signaled = 0;
+		nfc_tnep_events[1].state = K_POLL_STATE_NOT_READY;
+
+		tnep_states[current_state].action(TNEP_EVENT_TIMER_OFFSET
+					+ nfc_tnep_events[1].signal->result);
+
+		tnep_message_encode();
+	}
+
+	return 0;
+}
+
+int nfc_tnep_message_add_record_app_data(struct nfc_ndef_record_desc *record,
+					 bool is_last_app_record)
+{
+	if (!record) {
+		return -EINVAL;
+	};
+
+	int err = tnep_message_add_record(record);
+
+	if (is_last_app_record) {
```

Comment:
Shouldn't be timer stopped only after the NDEF message is encoded?

Reply:
Do we need to call ```nfc_tnep_new_message_indicator``` to trigger the NDEF message encoding after the descriptor is constructed?

Reply:
> Shouldn't be timer stopped only after the NDEF message is encoded?
We also need to stop timer when service is deselected.
It's good idea to move it form record adding to message encoding function.

Reply:
> Do we need to call `nfc_tnep_new_message_indicator` to trigger the NDEF message encoding after the descriptor is constructed?

Yes, we need it to generate signal for pooling mechanism.

Reply:
> We also need to stop timer when service is deselected.
It's good idea to move it form record adding to message encoding function.

Nice, then we don't this boolean parameter anymore in this function. You can indicate that message is comple calling ```nfc_tnep_new_message_indicator```

Reply:
We call `nfc_tnep_rx_msg_indicate` when new NDEF message from the Reader/Writer is available in buffor.
The boolean parameter in function adding application records informs program, there are no more record in tx message and it can be encoded intro buffer.
I think about rewriting code without this flag.